### ISSUE TITLE
Split `tsoa` package

### ISSUE THREAD

**Issue by drothmaler** [lukeautry/tsoa#546]
## Sorting

* **I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] support request

* I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit
    * There was #304 but without any real solution...

## Expected Behavior

Adding TSOA should not drastically increase the size of my deployed application (and it should not add unused dependencies to it).

## Current Behavior

Adding TSOA to a(n empty) project causes 65 npm packages to be added with a size of over 60MB (with 50MB for typescript alone). Especially having to include `typescript` in my deployed productive code is really bad in my opinion.

## Possible Solution

I think it would make much sense to split tsoa into 2 package, one for the CLI (`tsoa-cli`) and one Runtime package (`tsoa-runtime`).
This way `tsoa-cli` could be used as a global or dev dependency and only the needed code is deployed, with the productive runtime dependency. 
Runtime would probably just include: 
* controller base class
* decorators 
* and maybe a few other small things

## Breaking change?

Depending on the chosen package naming, the change might break existing builds and/or deployments (if they would be updated).

Possible options are: 

1. `tsoa`(CLI) --> `tsoa-runtime`
    * would probably not break anything
    * silent update might lead cause possible improvements to be wasted (as long as the CLI package stays in the productive dependencies)
2. `tsoa`(empty) --> `tsoa-cli` --> `tsoa-runtime`
    * basically the same as above, with nicer package name for CLI
3. `tsoa-cli` --> `tsoa` (Runtime)
   * would break the build, but not the runtime (as long as you don't try to actually use CLI features at runtime, but this should usually not be the case)
    * naming schema would be in line what many other (big) packages to (e.g. angular, babel, vue, gulp, grunt, karma, aurelia)
    * instant win for the runtime
    * fixing the build (by adding CLI package to the dev dependencies should be easy
4. `tsoa-cli` --> `tsoa-runtime`; deprecate `tsoa`
    * clear naming
    * would not break anything
    * forces explicit migration

I would prefer option 3 or 4 (event if they are slightly less comfortable), as 1 and 2 feel somewhat dirty to me.

cc: @ryo-nbx

**Comment by dgreene1** [comment]
I use tsoa as a devDependency so I don’t have this issue.

But I’ll cc other members of the tsoa community that might be able to better weigh in. @lukeautry @WoH

**Comment by drothmaler** [comment]
@dgreene1 but the compiles JavaScript files still use stuff from `tsoa`(e.g. decorators and base class for Controllers and the `ValidationService` inside the generated routes). 
So if you only install production dependencies (without `tsoa`) your app shouldn't work.

**Comment by dgreene1** [comment]
Oh maybe I don’t use tsoa as a dev dependency then. Sorry, I recently took a job doing C# so my knowledge of tsoa is fading.

But what I do recall is that I use webpack to reduce the size of my app when I’m running it in serverless (and where the size really matters due to cold start requirements). But if I’m using tsoa in a kubernetes / container hosted environment I rarely concern myself with the size of the app.

**Comment by dgreene1** [comment]
@drothmaler if you have experience with lerna then perhaps you'd like to introduce a PR that would allow us to deploy 3 npm packages?

Those 3 packages being:
* `@tsoa/cli`
* `@tsoa/runtime`
* `tsoa` (which is everything. Essentially for those like me who do not care about the package size)

But I do not have experience with lerna (outside of knowing that it's designed for this purpose). So you or someone who has the experience would have to take the initiative on this.

**Comment by drothmaler** [comment]
Well I played around with `lerna` for some time, but I haven't used it in a real project yet. But this seem like a good chance to get some practice :-)
I will provide a PR, as soon as I'm ready.

**Comment by marudor** [comment]
I would also love a package split.
In my experience I would not keep a single tsoa package and instead split it.
Keeping the all in one adds extra overhead and it's not that much of a hassle to install both.

**Comment by Kriskit** [comment]
I've submitted a PR for this as I am having issues with Webpack trying to bundle Typescript and other things that aren't needed for runtime and it adds quite a bit of size to the overall deployment.
